[ENHANCEMENT] Output build sizes command

martypenner opened the pull request:
This is a very naive implementation that simply checks the given asset path (`dist/assets` by default) for any `*.{css,js}` files and outputs their human-readable size. I chose to do it this way because of lack of access to the output tree when running this command.

The command itself is run like this: `ember show-asset-sizes`. If assets are stored in a non-standard directory, run `ember show-asset-sizes --asset-path path/to/assets`. The command name is a bit verbose, but I wasn't sure of a better name that would also convey the meaning.

Some notes:
- The command tests in `show-asset-sizes-test.js` are failing with "undefined is not a function". I cannot for the life of me seem to figure out what's going on there. Doing anything but `return`ing the command promise in the tests passes. As soon as I try to `return` the promise, the tests fail. My node-fu is weak.
- It does not run the `Builder` like the `serve` or `test` tasks do. I thought rebuilding the project might be frustrating when simply trying to see your asset sizes. If the asset directory can't be found, the task throws an error instead. If we decide to run the `Builder` in this task, then we could get rid of the cli argument for the asset path.

#4293

stefanpenner commented:
> The command tests in show-asset-sizes-test.js are failing with "undefined is not a function". I cannot for the life of me seem to figure out what's going on there. Doing anything but returning the command promise in the tests passes. As soon as I try to return the promise, the tests fail. My node-fu is weak.

I'll try to look sometime this week, remind me if i haven't by thursday

martypenner commented:
@stefanpenner Seems you like you've had a busy week. Any chance you were able to look at this?

stefanpenner commented:
Forgot to add a calendar reminder.  Got one set for this afternoon

stefanpenner commented:
one thing I noticed, it doesn't appear the build command calls the show-assets-sizes task

stefanpenner commented on lib/commands/show-asset-sizes.js:
```diff
@@ -0,0 +1,20 @@
+  description: 'Output asset file sizes.',
+
+  availableOptions: [
+    { name: 'asset-path', type: path, default: 'dist/assets' }
```
we may want this to come from --output-path, as users can configure this in there .emberrc

stefanpenner commented on tests/unit/commands/show-asset-sizes-test.js:
```diff
@@ -0,0 +1,78 @@
+
+  beforeEach(function () {
+    tasks = {
+      ShowSizes: Task.extend({
```
ShowAssetSizes

stefanpenner commented on tests/unit/commands/show-asset-sizes-test.js:
```diff
@@ -0,0 +1,78 @@
+      settings: {}
+    });
+
+    stub(tasks.ShowSizes.prototype, 'run');
```
ShowAssetSizes

stefanpenner commented on tests/unit/commands/show-asset-sizes-test.js:
```diff
@@ -0,0 +1,78 @@
+  });
+
+  afterEach(function () {
+    tasks.ShowSizes.prototype.run.restore();
```
ShowAssetSizes

stefanpenner commented on tests/unit/commands/show-asset-sizes-test.js:
```diff
@@ -0,0 +1,78 @@
+
+  it('has correct default value for asset path', function () {
+    return new ShowCommand(options).validateAndRun().then(function () {
+      var run = tasks.ShowSizes.prototype.run;
```
ShowAssetSizes

stefanpenner commented on tests/unit/commands/show-asset-sizes-test.js:
```diff
@@ -0,0 +1,78 @@
+    return new ShowCommand(options)
+      .validateAndRun(['--asset-path', 'some/path'])
+      .then(function () {
+        var run = tasks.ShowSizes.prototype.run;
```
ShowAssetSizes

stefanpenner commented on tests/unit/commands/show-asset-sizes-test.js:
```diff
@@ -0,0 +1,78 @@
+    return new ShowCommand(options)
+      .validateAndRun()
+      .then(function () {
+        var run = tasks.ShowSizes.prototype.run;
```
ShowAssetSizes

stefanpenner commented:
left some comments, i believe ShowSizes  -> ShowAssetSizes rename is trolling your tests currently

martypenner commented:
Wow, I was *convinced* that rename wasn't the issue. Good lesson in certainty vs. reality.

Re: the build command calling the show-asset-sizes task, that's now done. I wanted to wait until the test failures were figured out before adding complexity.

I'll keep working on this to increase code coverage and get travis and friends happy. I'll let you know once I think it's ready.

stefanpenner commented:
> I'll keep working on this to increase code coverage and get travis and friends happy. I'll let you know once I think it's ready.

let me know if I can help unblock you

martypenner commented:
BAM! I figured things out. Taught me a lot about JS testing too. Thanks for being so willing to help out.

I've changed the `--asset-path` command option to `--output-path` with an `o` alias, as you suggested. I wasn't sure how to write a test around providing options via .ember-cli, so I stubbed a test but marked it as `.skip`. I also provided a `--suppress-sizes` option to `build` which skips outputting file sizes. Not sure if that's necessary because of the globally available `--silent` option, so I'll defer to your judgement on that.

Any critiques welcome.

stefanpenner commented:
> BAM! I figured things out. Taught me a lot about JS testing too. Thanks for being so willing to help out.

thanks for pushing forward :)

@martypenner mind rebasing? I suspect these 40 commits can be coalesced into a few very clear commits.

stefanpenner commented on lib/commands/build.js:
```diff
@@ -22,7 +24,22 @@ module.exports = Command.extend({
+            outputPath: commandOptions.outputPath
+          });
+        }
+      }.bind(this));
```
i don't believe the `bind` is needed

stefanpenner commented on lib/models/asset-size-printer.js:
```diff
@@ -0,0 +1,59 @@
+          });
+      }.bind(this))
+      .catch(function (err) {
+        ui.writeLine(chalk.red(err));
```
typically we want the default error handler in ember-cli to handle this. So that: https://github.com/ember-cli/ember-cli/blob/master/lib/cli/cli.js#L150-L156 can print the errors nicely, and ensure the program exists with the right status code.

I believe you should just be able to omit the `catch` here.

martypenner commented:
I've pushed the concise commits. No idea why the [brocfile-smoke-test](https://travis-ci.org/ember-cli/ember-cli/jobs/72231533#L2622) is suddenly failing, though.

> i don't believe the bind is needed

Correct. I've removed it.

> typically we want the default error handler in ember-cli to handle this

Ah, didn't realize that. Fixed.

> mind rebasing? I suspect these 40 commits can be coalesced into a few very clear commits.

Is what I just pushed OK?

martypenner commented:
That test is failing on master locally as well.

stefanpenner commented:
> That test is failing on master locally as well.

Ah, I believe you need to prune `.node_modules-tmp` and `.bower_components-tmp`. Let me just make that automatic, as it trolls me often.

stefanpenner commented:
@martypenner sorry i haven't been much help as of late. I'll take a look tomorrow afternoon (cal event set)

stefanpenner commented:
> Ah, I believe you need to prune .node_modules-tmp and .bower_components-tmp. Let me just make that automatic, as it trolls me often. Although note, master should be green (CI claims it is)

https://github.com/ember-cli/ember-cli/pull/4522

chadhietala commented:
Restarting CI as #4522 is merged.

martypenner commented:
@stefanpenner No need to apologize. I imagine you've been super busy. I'm going to temporarily remove all my test changes to see if I can determine why the full test suite is failing.

martypenner commented:
Good thing for those acceptance tests. The brocfile-smoke-test was catching an incorrect assumption I made, but reporting it oddly. Fixed now. Unless there's other feedback, this should be good to merge.

martypenner commented:
Oops, fixed a typo. Should be good now.

stefanpenner commented:
Hey dude, one last request. can we also output the gzip'd sizes? (node ships with zlib, just be sure to use the async methods)

potential output:
```js
ember build --env production
version: 1.13.1-pr-4358-bdf52680a6
Built project successfully. Stored in "dist/".
File sizes:
vendor-7192b09eb86453dc45e12510994269d9.js: 695.76 kB (191.043kB gzip)
vendor-d41d8cd98f00b204e9800998ecf8427e.css: 0 B
with-divad-64f88b2bf9dff5730c32ecdb41b17822.js: 3.31 kB (   1.233 kB gzip)
with-divad-d41d8cd98f00b204e9800998ecf8427e.css: 0 B
```

rwjblue commented:
@stefanpenner - From what I can see, this integrates with a default `ember build`.  Seeing `gzip` output for `production` builds seems totally fine (since we are already in "bad build performance" land), but adding `gzip` to even development builds seems like it would needlessly slow things down when build speed should be the priority (during active development)...

stefanpenner commented:
> to even development builds seems like it would needlessly slow things down when build speed should be the priority (during active development)...

yes, but i don't believe this will slow down builds.

* this only happens when `ember build` is run
* gzip is insanely fast

rwjblue commented:
> one thing to note: I don't believe this will slow down builds.

Gotcha.

> this may only be valuable for production builds

Yes, I think that is the bigger issue. Displaying the build output sizes for non-minified (which is **definitely** slow) non-gzipped code seems not-terribly useful.

stefanpenner commented:
> Displaying the build output sizes for non-minified (which is definitely slow) non-gzipped code seems not-terribly useful.

i'll be pedantic, and claim this isn't slow:

```
time cat dist/assets/vendor-7192b09eb86453dc45e12510994269d9.js  | wc -c
        0.00 real         0.00 user         0.00 sy
```

rwjblue commented:
@stefanpenner - My words were phrased oddly, I was saying that minifying is slow.

stefanpenner commented:
> @stefanpenner - My words were phrased oddly, I was saying that minifying is slow.

english how does it work.

martypenner commented:
Sounds good to me. I can probably get this done tonight. End-of-week is doable for sure, barring any unforeseen learning curve.

stefanpenner commented:
> Sounds good to me. I can probably get this done tonight.

awesome

martypenner commented:
I got this done last night, but rebasing on master broke the build. I'll keep my eye open for a passing build today and rebase on that.

stefanpenner commented:
@martypenner this looks like a legit failure:

```

  1) models/asset-size-printer prints human-readable file sizes (including gzipped sizes) of css and js files in the output path:
     Error: Cannot find module 'filesize'
      at Class.Promise.resolve.then.then.files.forEach.sizeOutput (lib/models/asset-size-printer.js:9:620)
  2) models/asset-size-printer does not print gzipped file sizes of empty files:
     Error: Cannot find module 'filesize'
      at Class.Promise.resolve.then.then.files.forEach.sizeOutput (lib/models/asset-size-printer.js:9:620)
  3) models/asset-size-printer does not print project test helper file sizes:
     Error: Cannot find module 'filesize'
      at Class.Promise.resolve.then.then.files.forEach.sizeOutput (lib/models/asset-size-printer.js:9:620)
  4) models/asset-size-printer does not print non-css or js file sizes:
     Error: Cannot find module 'filesize'
      at Class.Promise.resolve.then.then.files.forEach.sizeOutput (lib/models/asset-size-printer.js:9:620)
  5) models/asset-size-printer prints an error when no files are found:
     AssertionError: expected 'Cannot find module \'filesize\'' to include 'No asset files found in the path provided: path/that/does/not/exist'
      at tests/unit/models/asset-size-printer-test.js:121:34
  6) Acceptance: smoke-test ember new foo, build production and verify fingerprint:
     the object {
  "code": 1
  "errors": []
  "output": [
    "Future versions of Ember CLI will not support v0.10.40. Please update to Node 0.12 or io.js.\n"
    "version: 1.13.1--7e22988c2b\n"
    "Could not find watchman, falling back to NodeWatcher for file system events.\n"
    "Visit http://www.ember-cli.com/user-guide/#watchman for more info.\n"
    "Building\n"
    "Built project successfully. Stored in \"dist/\".\n"
    "Cannot find module 'filesize'\n"
    "Error: Cannot find module 'filesize'\n    at Function.Module._resolveFilename (module.js:338:15)\n    at 
```

stefanpenner commented:
also im unsure why the badge on master is read, it appears to actually be green https://ci.appveyor.com/project/embercli/ember-cli

martypenner commented:
I. Am. So smart. S, M, R T. Ugh. Sorry. Travis is happy now, but appveyor is failing immediately.

stefanpenner commented:
@martypenner something is sideways on appveyor, likely unrelated to your PR. I will investigate this evening.

homu commented:
:umbrella: The latest upstream changes (presumably #4895) made this pull request unmergeable. Please resolve the merge conflicts.

martypenner commented:
Is there still interest in this feature? I'd like to clean this up if possible.

homu commented:
:umbrella: The latest upstream changes (presumably #4900) made this pull request unmergeable. Please resolve the merge conflicts.

stefanpenner commented:
> Is there still interest in this feature? I'd like to clean this up if possible.

@martypenner yes

martypenner commented:
Apologies for the delay. This is good to go on my end. The appveyor failures appear to be related to timeouts and directories not being removed in other tests.

homu commented:
:umbrella: The latest upstream changes (presumably #5213) made this pull request unmergeable. Please resolve the merge conflicts.

martypenner commented:
Conflicts resolved.

Turbo87 commented on lib/commands/build.js:
```diff
@@ -23,9 +25,23 @@ module.exports = Command.extend({
-    return win.checkWindowsElevation(this.ui).then(function() {
-      return buildTask.run(commandOptions);
+    return win.checkWindowsElevation(this.ui).then(function () {
+      return Promise.resolve()
```
why do you use nested promises here instead of `then()` chaining?

martypenner replied:
I recall that was because of `.then()` being undefined on `buildTask.run` within the tests. I misunderstood the usage of `stub`. I'm now providing `Promise.resolve()` as an argument to `stub`, so this should be resolved.

homu commented:
:umbrella: The latest upstream changes (presumably #5586) made this pull request unmergeable. Please resolve the merge conflicts.

martypenner commented:
Conflicts resolved.

homu commented:
:umbrella: The latest upstream changes (presumably #5584) made this pull request unmergeable. Please resolve the merge conflicts.

martypenner commented:
Fixed.

homu commented:
:umbrella: The latest upstream changes (presumably #5622) made this pull request unmergeable. Please resolve the merge conflicts.

stefanpenner commented:
rebase + small output change -> https://github.com/ember-cli/ember-cli/pull/5644

Will merge shortly, thanks dude. Sorry for lagging on this.
